### PR TITLE
feat(mep): subscript fix for multiple value arrays

### PR DESCRIPTION
**Pull request by onewland**
Functional change required: make `tags[key] = int_value` and `tags_raw[key] = text_value` work in queries
 
Mechanism:
1. change SubscriptableMapper to take a value subcolumn argument
2. change parsimonious special-casing around tag columns to accept `tags` or `tags_raw`

### CODE REVIEW

**Review by fpacifici — COMMENTED** [review]
Please see my comments inline

File: snuba/query/snql/parser.py [hunk 1]
```diff
@@ -185,7 +185,7 @@
     null_literal          = ~r"NULL"i
     subscriptable         = column_name open_square column_name close_square
     column_name           = ~r"[a-zA-Z_][a-zA-Z0-9_\.:]*"
-    tag_column            = "tags" open_square tag_name close_square
+    tag_column            = ("tags_raw"/"tags") open_square tag_name close_square
```
[the remark this answers is not in the document]
Comment (onewland) [reply snuba/query/snql/parser.py:188]: yeah, I didn't love this either but parsimonious was blowing up on (imo) valid snql queries until I added the special workaround

[the remark this answers is not in the document]
Comment (evanh) [reply snuba/query/snql/parser.py:188]: We delineate between tag and subscriptable columns, but they are both valid. You can see `subscriptable` as an expression two lines above this. That's how `context` and any other subscriptable should work. We treat `tags` queries differently from other subscriptable calls, which is why I created a separate expression for it.

[the remark this answers is not in the document]
Comment (onewland) [reply snuba/query/snql/parser.py:188]: I got this parse error:

```
2022-07-07 10:21:17,064 Invalid query
Traceback (most recent call last):
  File "/Users/oliver/workplace/snuba/./snuba/query/snql/parser.py", line 919, in parse_snql_query_initial
    exp_tree = snql_grammar.parse(body)
  File "/Users/oliver/workplace/snuba/.venv/lib/python3.8/site-packages/parsimonious/grammar.py", line 115, in parse
    return self.default_rule.parse(text, pos=pos)
  File "/Users/oliver/workplace/snuba/.venv/lib/python3.8/site-packages/parsimonious/expressions.py", line 122, in parse
    raise IncompleteParseError(text, node.end, self)
parsimonious.exceptions.IncompleteParseError: Rule 'query_exp' matched in its entirety, but it didn't consume all the text. The non-matching portion of the text begins with 'AND tags_raw[5] = 'h' (line 1, column 175).
```

for the following query:
```
MATCH (generic_metrics_distributions) SELECT min(value) AS total_seconds, arrayJoin(tags.raw_value) BY project_id, org_id, tags.raw_value WHERE org_id = 1 AND project_id = 2 AND tags_raw[5] = 'h' AND metric_id = 8 AND granularity = 2 AND timestamp >= toDateTime('2022-06-01T19:00:00') AND timestamp < toDateTime('2022-06-30T20:00:00')
```

and I could substitute any value for `tags_raw` and get the same error

[the remark this answers is not in the document]
Comment (evanh) [reply snuba/query/snql/parser.py:188]: I don't mind if you want to make the `column_name` in the inner brackets support a numeric value as well.

```
subscriptable         = column_name open_square subscriptable_key close_square
subscriptable_key     = column_name / integer_string
integer_string        = ~r"[0-9]+"
```

You'll have to add `visit_` functions for those as well.

**Review by evanh — COMMENTED** [review]
I would also write a test that uses the SDK to generate the SnQL query. I looked over the code and it shouldn't be a problem, but I would test it anyway.

File: snuba/clickhouse/translators/snuba/mappers.py [hunk 2]
```diff
@@ -135,6 +135,7 @@ class SubscriptableMapper(SubscriptableReferenceMapper):
     from_column_name: str
     to_nested_col_table: Optional[str]
     to_nested_col_name: str
+    value_subcolumn_name: str = "value"
```
[the remark this answers is not in the document]
Comment (onewland) [reply snuba/clickhouse/translators/snuba/mappers.py:138]: Verified that the only hardcoding of `.value` was in the file you pointed out (at least as far as uses of `mapping_pattern` is concerned) and fixed it.

File: snuba/query/snql/parser.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (onewland) [reply snuba/query/snql/parser.py:188]: > We delineate between tag and subscriptable columns, but they are both valid. You can see `subscriptable` as an expression two lines above this. That's how `context` and any other subscriptable should work. We treat `tags` expressions differently from other subscriptable calls, which is why I created a separate expression for it.
> 

> I don't mind if you want to make the `column_name` in the inner brackets support a numeric value as well.
> 
> ```
> subscriptable         = column_name open_square subscriptable_key close_square
> subscriptable_key     = column_name / integer_string
> integer_string        = ~r"[0-9]+"
> ```
> 
> You'll have to add `visit_` functions for those as well.

This is an extension of tags so unless we want to make subscript-able completely generic and to handle the tags case, I think I am taking the right approach here. Open to disagreement but I can also imagine that we might one day want to accept string literals or other arbitrary data as the key and then we'll just be redefining tags in subscriptable

[the remark this answers is not in the document]
Comment (evanh) [reply snuba/query/snql/parser.py:188]: Let's back up a bit, I think the point here is that what you're trying to do shouldn't require SnQL changes. You should be able to extend the the subscriptable mapper to pick the value column, and that's it.

Looking at this more, can you change `subscriptable` to take a `tag_column` value (maybe rename `tag_column` to something more generic)? Then you don't need to hardcode more column names into the SnQL language.

[the remark this answers is not in the document]
Comment (onewland) [reply snuba/query/snql/parser.py:188]: Is this preferable to you? (diff is based off HEAD of this branch, not master)

```
diff --git a/snuba/query/snql/parser.py b/snuba/query/snql/parser.py
index 8d03601c..a00d616f 100644
--- a/snuba/query/snql/parser.py
+++ b/snuba/query/snql/parser.py
@@ -183,9 +183,9 @@ snql_grammar = Grammar(
     true_literal          = ~r"TRUE"i
     false_literal         = ~r"FALSE"i
     null_literal          = ~r"NULL"i
-    subscriptable         = column_name open_square column_name close_square
+    subscriptable         = column_name open_square (column_name/tag_name) close_square
     column_name           = ~r"[a-zA-Z_][a-zA-Z0-9_\.:]*"
-    tag_column            = ("tags_raw"/"tags") open_square tag_name close_square
+    tag_column            = "tags" open_square tag_name close_square
     tag_name              = ~r"[^\[\]]*"
     identifier            = backtick ~r"[a-zA-Z_][a-zA-Z0-9_]*" backtick
     function_name         = ~r"[a-zA-Z_][a-zA-Z0-9_]*"
(END)
```

**Review by evanh — APPROVED** [review]
Thanks!

**Review by fpacifici — APPROVED** [review]

**Pull request merged**